### PR TITLE
Adds locale as a chainable operator to bar chart

### PR DESCRIPTION
**Pull request by maikdiepenbroek**
to enable passing a custom locale. This can be done in two forms: string (following the pattern: [a-z]{2}-[A-Z]{2} or an object in the form of a d3-format locale definition, f.e.: https://cdn.jsdelivr.net/npm/d3-format@1.4.2/locale/en-US.json. If a string is passed, it's locale is retrieved asynchronous and then passed to d3-format.formatDefaultLocale

### CODE REVIEW

File: src/charts/bar.js [hunk 1]
```diff
@@ -164,7 +166,16 @@ define(function(require) {
          * @param {BarChartData} _data The data to attach and generate the chart
          */
         function exports(_selection) {
-            _selection.each(function(_data) {
+            if (locale && enableLabels) {
+                setDefaultLocale(locale)
+                    .then(() => initChart(_selection))
```
Comment (DalerAsrorov) [inline src/charts/bar.js]: Can initialization of the locale be done somewhere else?

File: src/charts/helpers/locale.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+define(function (require) {
+    const {FORMAT_LOCALE_URL} = require('./constants');
+    const {formatDefaultLocale} = require('d3-format');
```
Comment (DalerAsrorov) [inline src/charts/helpers/locale.js]: There needs to be a test for this helper. Make sure that every use-case is covered.

File: src/charts/bar.js (hunk 1 above)
Comment (maikdiepenbroek) [reply src/charts/bar.js]: Sure, i first had this in the drawLabels call. What place suits you better ?

File: src/charts/helpers/locale.js (hunk 2 above)
Comment (maikdiepenbroek) [reply src/charts/helpers/locale.js]: Just to provide some context, this PR in its current state is only to discuss the current approach. If we decide it is the way to go, i can add it to the other charts and update docs/tests

**Comment by maikdiepenbroek** [comment]
@DalerAsrorov and @Golodhros As a side note, when i perform the linting script i get a lot of code-style changes, also i see a lot of unused variables coming from require etc. When this PR is in a later state, i'll create a separate commit that fixes linting issues etc.

File: src/charts/bar.js [hunk 3]
```diff
@@ -164,7 +166,16 @@ define(function(require) {
...
+            } else {
+                initChart(_selection);
+            }
+        }
+
+        function initChart(selection) {
+            selection.each(function(_data) {
```
Comment (Golodhros) [inline src/charts/bar.js]: I would say to trigger the locale request here and keep a reference to the promise. Something like:
```
  localeRequest = fetchLocale();
  selection.each(....
```
Later on, whenever the locale is actually needed, we'll do:
```
  localeRequest.then(...
```

That way we'll trigger the request as soon as possible (if needed) while blocking the render of anything that depends on it. 

We should probably include a configurable timeout in case the request fails.

Comment (maikdiepenbroek) [reply src/charts/bar.js]: Liked your suggestion, i've implemented it like this.

**Comment by maikdiepenbroek** [comment]
@Golodhros and @DalerAsrorov  If you've got the time please review the current state of this PR. So far i've added jsdoc and unit tests for the locale helper. Also i added a small unit test for the locale property in the bart chart. Next to that i've implemented the suggestion of storing a reference to the getLocale promise. 
I'm having a difficult time getting to know the way of testing the complete bar chart, maybe you could help me with rounding of the tests especially mocking/spying. 

Once the bar chart is tested to standards,  i can move on and implement the locale property in the other charts (where applicable) and expand the docs.

Sounds like a plan ?

File: src/charts/bar.js [hunk 4]
```diff
@@ -9,13 +9,13 @@ define(function(require) {
     const d3Format = require('d3-format');
     const d3Scale = require('d3-scale');
     const d3Selection = require('d3-selection');
-    const d3Transition = require('d3-transition');
```
Comment (Golodhros) [inline src/charts/bar.js]: This needs to be loaded, even if we don't use it.

File: src/charts/bar.js [hunk 5]
```diff
@@ -1292,6 +1302,15 @@ define(function(require) {
             return this;
         };
```
Comment (Golodhros) [inline src/charts/bar.js]: We would need a comment on this accessor so we generate the documentation

File: src/charts/bar.js [hunk 6]
```diff
@@ -177,7 +183,11 @@ define(function(require) {
                 drawBars();
                 drawAxis();
 
-                if (enableLabels) {
+                if (enableLabels && locale) {
+                    localeRequest
```
Comment (Golodhros) [inline src/charts/bar.js]: Would it work if we set up localeRequest as a resolved promise by default?

File: src/charts/helpers/locale.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    const { formatDefaultLocale } = require('d3-format');
+    const { json } = require('d3');
+    /**
+     * Sets the given locale as the new default locale through d3-format's formatDefaultLocale
+     * When a string is given as a parameter, it uses d3.json to fetch the correct locale definition.
+     * When an object is used, it simply uses it to set the new locale.
+     * @param  {string | Object} locale The locale in string or object format.
+     * @return {Promise}       Result of formatDefaultLocale call.
+     */
+
```
Comment (Golodhros) [inline src/charts/helpers/locale.js]: Nit: No space here, thanks!

File: src/charts/helpers/locale.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * Checks if a locale is of the correct format
+     * @param  {string}  locale Locale to check
+     * @return {Boolean}       Is the locale has a correct format (https://cdn.jsdelivr.net/npm/d3-format/locale/)
+     */
+
+    function isValidLocale(locale) {
+        return /[a-z]{2}-[A-Z]{2}/.test(locale);
```
Comment (Golodhros) [inline src/charts/helpers/locale.js]: I would move the RegEx as a constant at the top of this file.

File: src/charts/helpers/locale.js [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    function isValidLocaleDefinition(locale) {
+        const requiredKeys = ['decimal', 'thousands', 'grouping', 'currency']
```
Comment (Golodhros) [inline src/charts/helpers/locale.js]: I would move it to a constant at the top of the file as well.

File: src/charts/helpers/locale.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+     * Checks if a locale is of the correct format
+     * @param  {string}  locale Locale to check
+     * @return {Boolean}       Is the locale has a correct format (https://cdn.jsdelivr.net/npm/d3-format/locale/)
+     */
+
+    function isValidLocale(locale) {
+        return /[a-z]{2}-[A-Z]{2}/.test(locale);
+    }
+
+    function isValidLocaleDefinition(locale) {
```
Comment (Golodhros) [inline src/charts/helpers/locale.js]: Missing a small comment here, thanks!

File: test/specs/bar.spec.js [hunk 11]
```diff
@@ -767,6 +767,18 @@ define(['d3', 'bar', 'barChartDataBuilder'], function(d3, chart, dataBuilder) {
...
+
+                barChart.locale(testLocale);
+                newLocale = barChart.locale();
+
+                expect(defaultLocale).not.toBe(newLocale);
+                expect(newLocale).toBe(testLocale);
+            })
         });
     });
 });
```
Comment (Golodhros) [inline test/specs/bar.spec.js]: I would add another test within the 'render' section, checking that a label is different when using a different locale.

File: test/specs/helpers.spec.js [hunk 12]
```diff
@@ -441,5 +443,34 @@ define([
...
+                expectAsync(locale.setDefaultLocale(validLocaleDefinition)).toBeResolved();
+                done();
+            });
+
+            it('should return a rejected promise when a invalid locale definition is given', (done) => {
+                const invalidLocaleDefinition = {};
+                expectAsync(locale.setDefaultLocale(invalidLocaleDefinition)).toBeRejected();
+                done();
+            });
+        });
```
Comment (Golodhros) [inline test/specs/helpers.spec.js]: I would add a couple of test cases for the two OK cases.

File: package.json [hunk 13]
```diff
@@ -80,6 +80,7 @@
     "babel-eslint": "7.2.3",
     "babel-istanbul": "^0.12.2",
     "babel-loader": "7.1.2",
+    "babel-polyfill": "6.26.0",
```
Comment (Golodhros) [inline package.json]: Why did you need this? Only for the tests?

File: src/charts/bar.js (hunk 4 above)
Comment (maikdiepenbroek) [reply src/charts/bar.js]: Got it, i reversed it.

File: src/charts/bar.js (hunk 5 above)
Comment (maikdiepenbroek) [reply src/charts/bar.js]: Added it

File: src/charts/helpers/locale.js (hunk 7 above)
Comment (maikdiepenbroek) [reply src/charts/helpers/locale.js]: Removed it.

File: src/charts/helpers/locale.js (hunk 8 above)
Comment (maikdiepenbroek) [reply src/charts/helpers/locale.js]: Moved it to a const.

File: src/charts/helpers/locale.js (hunk 9 above)
Comment (maikdiepenbroek) [reply src/charts/helpers/locale.js]: Moved it to a const

File: src/charts/helpers/locale.js (hunk 10 above)
Comment (maikdiepenbroek) [reply src/charts/helpers/locale.js]: Added it

File: test/specs/helpers.spec.js (hunk 12 above)
Comment (maikdiepenbroek) [reply test/specs/helpers.spec.js]: I had those tests, but since i wasn't able to mock the d3.json call, i removed those tests since i would be relying on an internet connection to make the test pass. If that's not a problem, i could re add the OK cases test.

File: package.json (hunk 13 above)
Comment (maikdiepenbroek) [reply package.json]: This was required for the tests indeed, the parsing failed due to missing Promise support.

File: src/charts/bar.js (hunk 6 above)
Comment (maikdiepenbroek) [reply src/charts/bar.js]: I'm not sure what you mean, could you explain with some (pseudo) code ?

File: test/specs/bar.spec.js (hunk 11 above)
Comment (maikdiepenbroek) [reply test/specs/bar.spec.js]: I'm having trouble getting the test to pass, a little help here would be very much appreciated.

File: src/charts/bar.js (hunk 4 above)
Comment (Golodhros) [reply src/charts/bar.js]: Did you push the changes?

File: src/charts/bar.js (hunk 6 above)
Comment (Golodhros) [reply src/charts/bar.js]: Instead of declaring localeRequest as undefined, you could do:
  localeRequest = Promise.resolve();

Then, you wouldn't need to add the branching here:

   localeRequest.then(() => drawLabels())
         .catch(() => new Error('You need an Internet connection in order to load the selected locale'))

File: test/specs/bar.spec.js (hunk 11 above)
Comment (Golodhros) [reply test/specs/bar.spec.js]: OK, could you upload the test in a commented state?

File: test/specs/helpers.spec.js (hunk 12 above)
Comment (Golodhros) [reply test/specs/helpers.spec.js]: Makes sense, add those here commented for now, I will take a look later.

File: src/charts/bar.js (hunk 4 above)
Comment (maikdiepenbroek) [reply src/charts/bar.js]: No, did the changes locally only.

File: test/specs/bar.spec.js (hunk 11 above)
Comment (maikdiepenbroek) [reply test/specs/bar.spec.js]: Sure, coming up!

File: test/specs/helpers.spec.js (hunk 12 above)
Comment (maikdiepenbroek) [reply test/specs/helpers.spec.js]: Ok, re-added them

File: src/charts/bar.js (hunk 6 above)
Comment (maikdiepenbroek) [reply src/charts/bar.js]: I've added the Promise.resolve to the declaration of localeRequest, but i don't quite understand the  branching part.

Comment (Golodhros) [reply src/charts/bar.js]: If the value of localeRequest is a resolved promise by default, then you don't need to check for 'enableLables' or 'locale', you can just put the same code for all cases.

Comment (maikdiepenbroek) [reply src/charts/bar.js]: > If the value of localeRequest is a resolved promise by default, then you don't need to check for 'enableLables' or 'locale', you can just put the same code for all cases.

Got it now, good catch!

File: test/specs/bar.spec.js (hunk 11 above)
Comment (maikdiepenbroek) [reply test/specs/bar.spec.js]: I've added the commented test, been driving me crazy haha. I hope you can take a look. @Golodhros

File: src/charts/bar.js [hunk 14]
```diff
@@ -1292,6 +1300,22 @@ define(function(require) {
             return this;
         };
 
+        /**
+         * Gets or Sets the locale which d3-format uses.
+         * (Default is en-US)
+         * @param  {string} _x                   Desired locale
```
Comment (Golodhros) [inline src/charts/bar.js]: You can do [_x = 'es-US'] to mark the default.

Could you add a link in the comment to point to the possible values?
Also, if we accept a locale object instead of the string, we should explain its values or add another link for referencing the format of it.

File: src/charts/bar.js [hunk 15]
```diff
@@ -178,7 +184,9 @@ define(function(require) {
                 drawAxis();
 
                 if (enableLabels) {
-                    drawLabels();
+                    localeRequest
+                        .then(() => drawLabels())
```
Comment (Golodhros) [inline src/charts/bar.js]: Wouldn't the locale change also the axis tick text?

Comment (maikdiepenbroek) [reply src/charts/bar.js]: I'm not sure to be honest.

File: test/specs/bar.spec.js [hunk 16]
```diff
@@ -176,6 +176,44 @@ define(['d3', 'bar', 'barChartDataBuilder'], function(d3, chart, dataBuilder) {
                     expect(actualOffset).toBe(expectedOffset);
                 });
             });
+
+            // describe('Locale', () => {
```
Comment (Golodhros) [inline test/specs/bar.spec.js]: Nit: you can just add an 'x' as in 'xdescribe' to comment the whole thing.

Comment (maikdiepenbroek) [reply test/specs/bar.spec.js]: True, but we agreed on pushing a commented out test for you to have a look at, right ?

Comment (Golodhros) [reply test/specs/bar.spec.js]: Yeah, I mean that adding the 'x' is how you comment the test. Not a big deal though.

File: src/charts/bar.js (hunk 15 above)
Comment (Golodhros) [reply src/charts/bar.js]: No prob, I will test it later.

File: src/charts/bar.js (hunk 14 above)
Comment (maikdiepenbroek) [reply src/charts/bar.js]: Fixed

File: test/specs/bar.spec.js (hunk 16 above)
Comment (maikdiepenbroek) [reply test/specs/bar.spec.js]: Fixed

File: src/charts/bar.js [hunk 17]
```diff
@@ -210,28 +218,28 @@ define(function(require) {
         function buildContainerGroups() {
             let container = svg
                 .append('g')
-                  .classed('container-group', true)
-                  .attr('transform', `translate(${margin.left + yAxisPaddingBetweenChart}, ${margin.top})`);
+                .classed('container-group', true)
```
Comment (Golodhros) [inline src/charts/bar.js]: Oh, what happened here?
append is a special case that gets 2 spaces instead of four.
Was this eslint?

Comment (maikdiepenbroek) [reply src/charts/bar.js]: This happened after i ran `yarn lint:fix`

Comment (Golodhros) [reply src/charts/bar.js]: I see, is there an easy way of rollback that change only? It messed up all the indentation.

Comment (maikdiepenbroek) [reply src/charts/bar.js]: Rolled back before the linting 👍

**Comment by Golodhros** [comment]
Awesome work @maikdiepenbroek!!!

Let me play a bit with it and see how we move forward, thanks a lot!

**Comment by maikdiepenbroek** [comment]
@Golodhros No problem, been fun and thanks for the patience and collaboration 👍

**Comment by Golodhros** [comment]
> @Golodhros No problem, been fun and thanks for the patience and collaboration 👍

On the contrary! Thanks for your patience and sorry for the back and forths!
